specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1086 @@
+    // Type of Consensus algorithm.
+    Consensus hierarchical.ConsensusType
+    // Minimum stake required for an address to join the subnet
+    // as a miner
+    MinMinerStake abi.TokenAmount
+    // List of miners in the subnet.
+    // NOTE: Consider using AMT.
+    Miners []address.Address
+    // Total collateral currently deposited in the
+    TotalStake abi.TokenAmount
+    // BalanceTable with the distribution of stake by address
+    Stake cid.Cid // HAMT[tokenAmount]address
+    // State of the subnet (Active, Inactive, Terminating)
+    Status Status
+    // Genesis bootstrap for the subnet. This is created
+    // when the subnet is generated.
+    Genesis []byte
+    // Checkpointing period. Number of epochs between checkpoint commitments
+    CheckPeriod abi.ChainEpoch
+    // Checkpoints submit to SubnetActor per epoch
+    Checkpoints cid.Cid // HAMT[epoch]Checkpoint
+    // CheckpointVotes includes the validator votes for the checkpoint
+    // of the current window.
+    CheckpointVotes cid.Cid // HAMT[cid]CheckVotes
+    // ValidatorSet is a set of validators
+    ValidatorSet []hierarchical.Validator
+    // MinValidators is the minimal number of validators required to join before starting the subnet
+    MinValidators uint64
+}
+```
+##### Constructor Parameters
+> Constructor Parameters for reference implementation of subnet actor.
+```go
+type ConstructParams struct {
+    // ID of the current network
+    NetworkName     address.SubnetID      
+    // Human-readable name for the subnet
+    Name            address.SubnetID  
+    // Consensus implemented in the subnet
+    Consensus hierarchical.ConsensusType
+    // Arbitrary consensus parameters required to initialize
+    // the consensus protocol (e.g. minimum number of validators).
+    // These parameters are consensus-specific.
+    ConsensusParams *hierarchical.ConsensusParams
+    // Minimum amount of stake required from participant 
+    // to become a validator of the subnet.
+    MinValidatorStake   abi.TokenAmount
+    // Checkpointing period used in the subnet. It determines
+    // how often the subnet propagates and commits checkpoints in its
+    // parent
+    CheckPeriod     abi.ChainEpoch
+}
+```
+> TODO: There is an on-going revision of `ConstructParams` to include a way to provide arbitrary input arguments to a subnet.
+
+## Subnet Coordinator Actor (SCA)
+The main entity responsible for handling all the lifecycle of child subnets in a specific chain is the Subnet Coordinator Actor (`SCA`). The `SCA` is a builtin-actor that exposes the interface for subnets to interact with the hierarchical consensus protocol. This actor includes all the available functionalities related to subnets and their management. It also enforces all the security assumptions, fund management, and cryptoeconomics of the hierarchical consensus, as Subnet Actors are user-defined and can't be (fully) trusted. The `SCA` has a reserved address ID `f064`.
```

Comment:
```suggestion
The main entity responsible for handling all the lifecycle of child subnets in a specific chain is the Subnet Coordinator Actor (`SCA`). The `SCA` is a builtin-actor that exposes the interface for subnets to interact with the hierarchical consensus protocol. This actor includes all the available functionalities related to subnets and their management. It also enforces all the security requirements, fund management, and cryptoeconomics of the hierarchical consensus, as Subnet Actors are user-defined and can't be (fully) trusted. The `SCA` has a reserved address ID `f064`.
```

Enforcing an assumption sounds strange to me, but feel free to keep it. Intuitively, if one could enforce it, it wouldn't be an assumption.